File: ru.kmorozov.library.data.model/src/main/java/ru/kmorozov/library/data/model/dto/ItemDTO.java
package ru.kmorozov.library.data.model.dto;

import org.springframework.hateoas.ResourceSupport;

import java.util.List;
import java.util.stream.Collectors;

/**
 * Created by sbt-morozov-kv on 31.01.2017.
 */
public class ItemDTO extends ResourceSupport {

    public static final int REFRESH_INTERVAL = 10 * 1000 * 60;

    public enum ItemType {
        storage,
        category,
        book
    }

    public enum RefreshStatus {
        dirty,
        updated
    }

    private String itemId;
    private ItemDTO.ItemType itemType;
    private Object itemSubType;
    private String displayName;
    private long filesCount;
    private List<CategoryDTO> categories;
    private ItemDTO.RefreshStatus refreshStatus;

    public ItemDTO() {
    }

    public ItemDTO(StorageDTO storageDTO) {
        itemId = storageDTO.getId();
        itemType = ItemDTO.ItemType.storage;
        itemSubType = storageDTO.getStorageType();
        displayName = storageDTO.getDisplayName();
        filesCount = storageDTO.getFilesCount();

        this.refreshStatus = (long) REFRESH_INTERVAL > System.currentTimeMillis() - storageDTO.getLastChecked() ? ItemDTO.RefreshStatus.updated : ItemDTO.RefreshStatus.dirty;

        if (null != storageDTO.getCategories())
            categories = storageDTO.getCategories().stream().map(CategoryDTO::new).collect(Collectors.toList());
    }

    public ItemDTO(BookDTO bookDTO) {
        itemId = bookDTO.getId();
        itemType = ItemDTO.ItemType.book;
        itemSubType = bookDTO.getFormat();
        displayName = bookDTO.getTitle();
    }

    public String getItemId() {
        return this.itemId;
    }

    public ItemDTO.ItemType getItemType() {
        return this.itemType;
    }

    public Object getItemSubType() {
        return this.itemSubType;
    }

    public String getDisplayName() {
        return this.displayName;
    }

    public long getFilesCount() {
        return this.filesCount;
    }

    public List<CategoryDTO> getCategories() {
        return this.categories;
    }

    public ItemDTO.RefreshStatus getRefreshStatus() {
        return this.refreshStatus;
    }

    public void setUpdated() {
        refreshStatus = ItemDTO.RefreshStatus.updated;
    }
}

File: ru.kmorozov.library.data.model/src/main/java/ru/kmorozov/library/data/model/dto/ItemDTO.kt
package ru.kmorozov.library.data.model.dto

import kotlin.streams.toList
import org.springframework.hateoas.RepresentationModel

/**
 * Created by sbt-morozov-kv on 31.01.2017.
 */
class ItemDTO : RepresentationModel<ItemDTO> {

    lateinit var itemId: String
        private set
    lateinit var itemType: ItemType
        private set
    lateinit var itemSubType: Any
        private set
    lateinit var displayName: String
        private set
    var filesCount: Long = 0
    var categories: List<CategoryDTO>? = null
        private set
    var refreshStatus: RefreshStatus? = null
        private set

    enum class ItemType {
        storage,
        category,
        book
    }

    enum class RefreshStatus {
        dirty,
        updated
    }

    constructor()

    constructor(storageDTO: StorageDTO) {
        this.itemId = storageDTO.id
        this.itemType = ItemType.storage
        this.itemSubType = storageDTO.storageType
        this.displayName = storageDTO.displayName
        this.filesCount = storageDTO.filesCount

        refreshStatus = if (REFRESH_INTERVAL.toLong() > System.currentTimeMillis() - storageDTO.lastChecked) RefreshStatus.updated else RefreshStatus.dirty

        if (null != storageDTO.categories)
            this.categories = storageDTO.categories!!.stream().map { CategoryDTO(it) }.toList()
    }

    constructor(bookDTO: BookDTO) {
        this.itemId = bookDTO.id
        this.itemType = ItemType.book
        this.itemSubType = bookDTO.format
        this.displayName = bookDTO.title
    }

    fun setUpdated() {
        this.refreshStatus = RefreshStatus.updated
    }

    companion object {

        const val REFRESH_INTERVAL = 10 * 1000 * 60
    }
}
